Document byte_update_exprt and rename non-const functions

romainbrenguier opened the pull request:
In some cases the non-const version was unnecessarily called because it
has the same name as the const version.
This pull requests avoids that from happening in 9 cases for op, 2 for value and 3 for
offset.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Document byte_extract_exprt

Documentation was missing

romainbrenguier committed:
Document byte_update_exprt

The documentation was missing.

romainbrenguier committed:
Define set_* non-const byte_update_exprt accessors

In some cases the non-const version was unnecessarily called because it
has the same name as the const version.
We define accessor functions with a different name to make the
non-constantness explicit.

tautschnig reviewed (changes requested):
Renaming functions without following an appropriate deprecation path is a no-go.

romainbrenguier committed:
Deprecate non-const byte_update_exprt accessors

The non-const accessors which have the same name as the const one
shouldn't be used anymore to avoid confusion and unknowningly calling
the non-const version.

romainbrenguier committed:
Remove use of deprecated non-const accessors

The non-const version shouldn't be used anymore when we don't actually
need to modify the operands.

romainbrenguier commented:
@tautschnig 
> Renaming functions without following an appropriate deprecation path is a no-go.

I've now kept both versions and deprecated the old one.

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -2057,20 +2057,22 @@ bool simplify_exprt::simplify_byte_extract(byte_extract_exprt &expr)
 
 bool simplify_exprt::simplify_byte_update(byte_update_exprt &expr)
 {
+  const byte_update_exprt &expr_const = as_const(expr);
```
Is this use of `as_const` actually necessary?

tautschnig reviewed (approved):
Looks good to me - except I'd prefer if the `set_*` methods actually took an argument as discussed.
